File: python/1457. Pseudo-Palindromic Paths in a Binary Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
from pyparsing import Optional


class Solution:
    def pseudoPalindromicPaths (self, root: Optional[TreeNode]) -> int:
        return self.dpHelper(root, set())
    def dpHelper(self, root, frequencySet):
        if root == None:
            return 0
        else:
            frequencySetCopy = frequencySet.copy()
            if root.val in frequencySet:
                frequencySetCopy.remove(root.val)
            else:
                frequencySetCopy.add(root.val)
            if (root.left == None and root.right == None):
                return 1 if len(frequencySetCopy) <= 1 else 0
            else:
                return self.dpHelper(root.left, frequencySetCopy) + self.dpHelper(root.right, frequencySetCopy)

File: java/1457. Pseudo-Palindromic Paths in a Binary Tree.java

// Definition for a binary tree node.

import java.util.HashSet;

class Solution {
       public class TreeNode {
       int val;
       TreeNode left;
       TreeNode right;
       TreeNode() {}
       TreeNode(int val) { this.val = val; }
       TreeNode(int val, TreeNode left, TreeNode right) {
           this.val = val;
           this.left = left;
           this.right = right;
       }
   }
    public int pseudoPalindromicPaths (TreeNode root) {
        return helper(root, new HashSet<Integer>());
    }
    public int helper(TreeNode root, HashSet<Integer> hs){
        if(root == null){
            return 0;
        }
        HashSet<Integer> hsCopy = new HashSet<Integer>(hs);
        if(hs.contains(root.val)){
            hsCopy.remove(root.val);
        }
        else{
            hsCopy.add(root.val);
        }
        if(root.left == null && root.right == null){
        if(hsCopy.size() > 1){
            return 0;
        } else {
            return 1;
        }
        }
        return helper(root.left,hsCopy) + helper(root.right,hsCopy);
    }

}